File: HW9/HW9.py

class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):

        self.val = val

        self.left = left

        self.right = right
    
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':

            return LCA_helper(root,p,q)

     


     
def LCA_helper( root, p, q):
    if not root:
        return None

    elif root.val == p.val or root.val == q.val:
        return root

    leftSubTree = LCA_helper(root.left, p, q)
    rightSubTree = LCA_helper(root.right, p, q)

    if leftSubTree and rightSubTree:
        return root

    elif not leftSubTree:
        return rightSubTree

    elif not rightSubTree:
        return leftSubTree

    return None
     
     
        
def printNode(root):
     if root is not None:
        print(root.val)

if __name__ == "__main__":
    p = TreeNode(3)
    q = TreeNode(1)
    r = TreeNode(5)

    root = TreeNode(4)
    root.left =  TreeNode(3)
    root.right =  TreeNode(8)
    root.left.left =  TreeNode(1)
    root.right.left =  TreeNode(5)
    root.right.right =  TreeNode(9)
    thy_node = root.lowestCommonAncestor(root, p, q)
    thy_node2 = root.lowestCommonAncestor(root, q, r)

    print(f"here is the expected LCA of {p.val}  and {q.val}: ")
    printNode(thy_node)
    print("\n")
    print(f"here is the expected LCA of {q.val} and {r.val}: ")
    printNode(thy_node2)

File: HW9/HW9.java
public class HW9 {

    
    public static TreeNode LCA_Helper(TreeNode root, TreeNode p, TreeNode q){
        if(root == null){
            return null;
        }
        else if(root.val == q.val || root.val == p.val){
            return root;

        }

            TreeNode leftSubTree = LCA_Helper(root.left, p, q);
            TreeNode RightSubTree = LCA_Helper(root.right, p, q);


        if(leftSubTree !=null && RightSubTree != null){

            return root;
        }
        else if( leftSubTree == null){
            return RightSubTree;
        }

        else if(RightSubTree == null){
            return leftSubTree;
        }

        return null;
        
    }

    
    
    
    
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        

        return LCA_Helper(root, p, q); 
     }

     
     

     public void printNode(TreeNode root, TreeNode p, TreeNode q) {
        TreeNode lca = lowestCommonAncestor(root, p, q);
        if (lca != null) {
            System.out.println("The LCA of " + p.val + " and " + q.val + " is: " + lca.val);
        } else {
            System.out.println("No LCA found for " + p.val + " and " + q.val);
        }
    }

     
     
     
     

     public static void main(String args[]){  
        HW9 hw9 = new HW9(); // god I hate java sometimes so much such a pointless operation of creating a class variable of a useless class just so i can have no errors god help me ima pop a blood vessel took me an hour to figure out the issue grrr.
        TreeNode p = new TreeNode(3);
        TreeNode q = new TreeNode(1);
        TreeNode r = new TreeNode(5);

        TreeNode root = new TreeNode(4);
        root.left = new TreeNode(3);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(1);
        root.right.left = new TreeNode(5);
        root.right.right = new TreeNode(9);

        hw9.printNode(root,p,q);
        hw9.printNode(root,p,r);
    }
}
